crt/debugmalloc.c

```diff
@@ -0,0 +1,538 @@
+// Copyright (c) Open Enclave SDK contributors.
+// Licensed under the MIT License.
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
+#include <stdlib.h>
+
+#include <myst/spinlock.h>
+#include <myst/mmanutils.h>
+
+#define BACKTRACE_MAX 16
+
+#define ENABLE_MALLOC_MEMSET
+#define ENABLE_FREE_MEMSET
+
+#define RETURN_ADDRESS(nr) \
+  __builtin_extract_return_addr(__builtin_return_address (nr))
+
+extern bool __crt_enable_debug_malloc;
+
+/* weak musl libc functions used by this module */
+extern void* libc_malloc(size_t size);
+extern void libc_free(void* ptr);
+extern void* libc_memalign(size_t alignment, size_t size);
+extern void* libc_calloc(size_t nmemb, size_t size);
+extern void* libc_realloc(void* ptr, size_t size);
+extern size_t libc_malloc_usable_size(void* ptr);
+
+MYST_NOINLINE
+size_t myst_backtrace_impl(void** start_frame, void** buffer, size_t size)
```

Comment:
This assumes that the application is not compiled with `-fomit-frame-pointer`. Maybe we should surface this in the documentation, so that users can be aware and build accordingly?